File: 섹션03. 2차원 배열 회전하기/05. BOJ 1913 - 달팽이.cpp
#include <stdio.h>

#define MAX (1000 + 100)

int N, K;
int MAP[MAX][MAX];

// 상, 우, 하, 좌 (↑, →, ↓, →)
int dr[] = { -1, 0, 1, 0 };
int dc[] = { 0, 1, 0, -1 };

void printMap()
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
			printf("%d ", MAP[r][c]);
		putchar('\n');
	}
}

void makeSnail()
{
	int ansR, ansC;
	int r, c, direction;
	int index, size;
		
	r = c = ansR = ansC = (N + 1) / 2;
	direction = 0;
	index = 1;
	size = 0;

	MAP[r][c] = index++;

	for (int i = 0; i < 2 * N - 1; i++)
	{
		if(i % 2 == 0) size++;

		for (int s = 0; s < size; s++)
		{
			int nr, nc;

			nr = r + dr[direction];
			nc = c + dc[direction];

			if (index == K)
			{
				ansR = nr;
				ansC = nc;
			}

			MAP[nr][nc] = index++;

			r = nr;
			c = nc;
		}

		direction++;

		if (direction == 4) direction = 0;
	}

	printMap();

	printf("%d %d\n", ansR, ansC);
}

int main()
{
	scanf("%d %d", &N, &K);

	makeSnail();

	return 0;
}

File: Python/섹션03. 2차원 배열 회전하기/05. BOJ 1913 - 달팽이.py
MAX = 1000 + 100

dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

MAP = [[0] * MAX for _ in range(MAX)]

N = int(input())
K = int(input())

def print_map():
    for r in range(1, N + 1):
        print(*MAP[r][1:N + 1])
    print("")

def make_snail():
    r = c = ansR = ansC = (N + 1) // 2
    direction = 0
    index = 1
    size = 0
    
    MAP[r][c] = index
    index += 1
    
    for i in range(2 * N - 1):
        if i % 2 == 0: size += 1
        
        for s in range(size):
            nr = r + dr[direction]
            nc = c + dc[direction]
            
            if index == K : ansR, ansC = nr, nc
            
            MAP[nr][nc] = index
            index += 1
            
            r, c = nr, nc
            
        direction += 1
        
        if direction == 4: direction = 0
    
    print_map()
    
    print(ansR, ansC)
    
make_snail()
